Зотов Андрей

anzotov opened the pull request.

anzotov committed:
Первая реализация

anzotov committed:
Minor fix

dsaa27 commented on .clang-format:
```diff
@@ -106,7 +106,6 @@ BraceWrapping:
     BeforeElse: true
     BeforeCatch: true
     BeforeLambdaBody: false
-    SplitEmptyNamespace: false
```
👍

dsaa27 commented on devicemonitoringserver/CMakeLists.txt:
```diff
@@ -8,5 +8,8 @@ FILE(GLOB SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/*.h ${CMAKE_CURRENT_SOURCE_DIR}/*.
 
 add_executable(${PROJECT_NAME} ${SOURCES})
 target_include_directories(${PROJECT_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
+set(GCC_OPTIONS -g -O0 -Werror -Wall -Wextra -fsanitize=leak -fsanitize=undefined -fsanitize=address)
```
👍

dsaa27 commented on devicemonitoringserver/baseencoderexecutor.h:
```diff
@@ -0,0 +1,27 @@
+    /*!
+     * \brief Расшифровать сообщение \a message
+     */
+    virtual std::string decode(const std::string& message) = 0;
```
Методы encode и decode можно сделать константными?

anzotov replied:
Я исходил из предположения, что алгоритму шифрования может понадобиться хранить данные внутри себя, поэтому не стал их делать константными. Хотя, если наша система проектируется для работы в многопоточном режиме, константность позволит снизить вероятность ошибок.

dsaa27 replied:
Неконстантность этих методов предполагала бы, что алгоритм шифрования учитывает прежние сообщения для кодирования/декодирования последующих. Подразумеваем, что таких алгоритмов мы использовать не будем (не знаю, есть ли такие), ведь тогда потеря по сети одного сообщения привела бы к невозможности декодировать последующие.
Но я кажется понял твою мысль.

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/bigendian.h:
```diff
@@ -0,0 +1,41 @@
+    {
+        buf[i] = value;
+        value >>= 8;
+    }
```
Кажется операция сдвига над знаковым целым является UB (пусть как правило он и реализуется как арифметический сдвиг), поэтому этот код вызывает неуверенность.
Предлагаю перед переводом скастить тип Т к аналогичному или большему по размеру беззнаковому.

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/bigendian.h:
```diff
@@ -0,0 +1,41 @@
+    }
+    for (size_t i = bytes_num; i > 0; --i)
+    {
+        ostream.put(buf[i - 1]);
```
можно попробовать заиспользовать std::copy + std::ostream_iterator

dsaa27 replied:
И вообще, можно же не заводить локальный массив, а сразу писать в поток, если сдвигать не само значение value, а маску 0xFF. Тогда и проблем со сдвигом знакового value не будет

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/bigendian.h:
```diff
@@ -0,0 +1,41 @@
+    T value {};
+    for (size_t i = 0; i < sizeof(T); ++i)
+    {
+        value <<= 8;
```
То же самое: такие операции лучше над беззнаковыми типами выполнять.
Предлагаю сдвигать не само значение value, а добавляемый в него через битовое ИЛИ байт

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/rot3encoderexecutor.cpp:
```diff
@@ -0,0 +1,40 @@
+std::string Rot3EncoderExecutor::encode(const std::string& message)
+{
+    std::string encoded(message);
+    for (auto& ch : encoded)
```
С т.з. использования кэша процессора гораздо эффективнее читать в цикле из одного, неизменяемого, массива (message) и писать в другой, результирующий, массив (encoded)

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/rot3encoderexecutor.h:
```diff
@@ -0,0 +1,14 @@
+
+#include "baseencoderexecutor.h"
+
+class Rot3EncoderExecutor : public BaseEncoderExecutor
```
final

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/rot3encoderexecutor.h:
```diff
@@ -0,0 +1,14 @@
+class Rot3EncoderExecutor : public BaseEncoderExecutor
+{
+public:
+    std::string encode(const std::string& message) override;
```
final

anzotov replied:
Если класс объявлен final, зачем на виртуальную функцию вешать final?

dsaa27 replied:
Ну, это не критично, просто ваш ревьюер является сторонником максимальных ограничений)
Если коллега снимет final с класса, то при наличии final на методах ему придется снимать их отдельно отдельно и заставит уделить большее внимание возможности переопределения каждого метода.

dsaa27 commented on devicemonitoringserver/rot3encoderexecutor.h:
```diff
@@ -0,0 +1,14 @@
+
+class Rot3EncoderExecutor : public BaseEncoderExecutor
+{
+public:
```
можно все в private спрятать, если полагаем, что должны взаимодействовать с реализациями только через интерфейс

anzotov replied:
Выполнено.

dsaa27 replied:
Это тоже из соображений максимальных ограничений

dsaa27 commented on devicemonitoringserver/messageserializer.h:
```diff
@@ -0,0 +1,28 @@
+    /*!
+     * \brief Сериализовать сообщение \a message
+     */
+    std::string serialize(const Message& message);
```
const ? Или даже static

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/message.h:
```diff
@@ -0,0 +1,30 @@
+    /*!
+    * \brief Вывод сообщения в виде строки для отладки
+    */
+    virtual std::ostream& print(std::ostream& os) const = 0;
```
нужно ли методу print возвращаемое значение?
Ведь при самостоятельном его использовании (не в рамках оператора вывода) им все равно вряд ли будешь пользоваться.

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/messageencoder.h:
```diff
@@ -0,0 +1,42 @@
+    /*!
+     * \brief Зашифровать сообщение \a message
+     */
+    std::string encode(const std::string& message);
```
const ?

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/messageserializer.cpp:
```diff
@@ -0,0 +1,102 @@
+            }
+            if (success)
+                break;
+        }
```
предлагаю логику сериализации/десериализации данных сообщения отдать самому классу сообщения, а здесь оставить только запись и чтение типа сообщения.

anzotov replied:
Выполнено.

dsaa27 replied:
Почти.
Можно еще было бы завести метод в базовом классе, который бы возвращал тип наследника

dsaa27 commented on devicemonitoringserver/messageserializer.cpp:
```diff
@@ -0,0 +1,102 @@
+    return ostream.str();
+}
+
+bool MessageSerializer::deserialize(const std::string& string, std::function<void(const Message&)> callback)
```
Если необходимо показать/учесть, что функция не всегда возврает результат, можно воспользоваться std::optional

anzotov replied:
Не понял, что мне заворачивать в std::optional? У меня результат - это успешность или неуспешность десериализации. Возвращать std::optional\<bool\> - это масло масляное какое-то :)

dsaa27 replied:
Имел в виду, что если у тебя функция возвращает какой-то результат вычислений и при этом не всегда, то часто вместо подхода "вернуть признак успеха через return и результат вычислений через аргумент" удобно использовать подход вернуть результат вычислений как std::optional. Ведь в этом случае у тебя данные и признак их валидности соединены в одном объекте, а не разнесены по двум разным переменным, и прочитать невалидные данные не получится.

dsaa27 replied:
Кроме того передача колбека для получения результата может ошибочно вводить читателя в заблуждение будто функция асинхронная.

anzotov replied:
Согласен, но коллбек тут используется по другой причине. Мне нужно было вернуть из функции объект неизвестного типа, унаследованный от класса Message. Нельзя создать объект на стеке внутри функции и вернуть его по ссылке/указателю на базовый класс, а создавать объект в куче не хотелось, т.к. в данном случае объект нужен только на короткое время, чтобы получить из него данные в пользовательском коде. Поэтому я решил сам передавать объект в пользовательскую функцию. Если есть более грамотное решение этой проблемы, хотелось бы узнать :)

dsaa27 commented on devicemonitoringserver/commandcenter.h:
```diff
@@ -0,0 +1,63 @@
+{
+    Phase phase;             ///< Этап плана
+    uint64_t firstTimestamp; ///< Временная метка первого измерения для данного этапа плана
+    double deviation;        ///< СКО физического параметра от плана
```
Встроенным типам надо дать какие-нибудь начальные значения, например 0 и NaN

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/commandcenter.h:
```diff
@@ -0,0 +1,63 @@
+
+class DeviceMonitoringServer;
+class MessageMeterage;
+class Message;
```
👍

dsaa27 commented on devicemonitoringserver/commandcenter.h:
```diff
@@ -0,0 +1,63 @@
+/*!
+ * \brief Класс командного центра для управления и ведения статистики по физическим параметрам
+ */
+class CommandCenter
```
final

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/commandcenter.h:
```diff
@@ -0,0 +1,63 @@
+    struct ScheduleInfo
+    {
+        std::vector<Phase> phases;
+        size_t currentPhaseIndex;
```
= 0 или = InvalidIndex (= std::numeric_limits<size_t>::max())

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/commandcenter.h:
```diff
@@ -0,0 +1,63 @@
+    /*!
+     * \brief Статистика СКО физических параметров от плана для устройства с идентификатором \a deviceId
+     */
+    std::vector<DeviationStats> deviationStats(uint64_t deviceId);
```
const

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/commandcenter.cpp:
```diff
@@ -0,0 +1,77 @@
+void CommandCenter::setSchedule(const DeviceWorkSchedule& schedule)
+{
+    m_scheduleInfo[schedule.deviceId] = {};
+    auto& scheduleInfo = m_scheduleInfo[schedule.deviceId];
```
Зачем дважды вести поиск по ключу?

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/commandcenter.cpp:
```diff
@@ -0,0 +1,77 @@
+
+void CommandCenter::setSchedule(const DeviceWorkSchedule& schedule)
+{
+    m_scheduleInfo[schedule.deviceId] = {};
```
Думаю, лучше явно currentPhaseIndex нулем проинициализировать, или как минимум явно задать дефолтное значение в самой структуре

anzotov replied:
Выполнено (задал дефолтное значение).

dsaa27 commented on devicemonitoringserver/commandcenter.cpp:
```diff
@@ -0,0 +1,77 @@
+        callback(deviceId, MessageError(MessageError::ErrorType::Obsolete));
+        return;
+    }
+    m_lastTimeStamp[deviceId] = currentTimeStamp;
```
Может лучше только для устройств, для которых задан план, запоминать какие-то параметры?

anzotov replied:
Я тут подумал про принцип Fail Fast: если устройство шлет сообщения с неверными временными метками, лучше ему сразу об этом сообщить, чем дожидаться выставления плана для этого устройства. Но если цель - экономия ресурсов сервера, можно, конечно, этого не делать.

dsaa27 replied:
Понял тебя

dsaa27 commented on devicemonitoringserver/devicemonitoringserver.h:
```diff
@@ -32,6 +35,14 @@ class DeviceMonitoringServer
+     * \brief Ссылка на объект MessageEncoder для управления параметрами шифрования.
+     */
+    MessageEncoder& messageEncoder();
 
```
Было бы неплохо еще добавить метод, который позволил бы нам забывать об устройствах, т.е. удалять данные о планах и статистике по устройству

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/tests.cpp:
```diff
@@ -1,27 +1,807 @@
-void monitoringServerTest1()
+#include <limits>
+
+#define compareVectorsOfSharedPtrs(a, b)  \
```
Как правило лучше чтобы макрос не мимикрировал по написанию под функцию: COMPARE_VECTORS_OF_SHARED_PTRS

anzotov replied:
Выполнено.

dsaa27 commented on devicemonitoringserver/devicemock.h:
```diff
@@ -71,6 +84,9 @@ class DeviceMock
     AbstractClientConnection* m_clientConnection = nullptr;
     std::vector<uint8_t> m_meterages;
     uint64_t m_timeStamp = 0;
+    std::vector<std::shared_ptr<Message>> m_messages;
```
Правда нужен shared_ptr, а не unique_ptr ?

anzotov replied:
С unique_ptr код не компилируется. Вероятно, связано с тем, что List-initialization копирует элементы, а не перемещает. Видимо, чтобы все заработало через unique_ptr, нужно применять [такой подход](https://godbolt.org/z/9nrxfWP6Y). Но проще оставить shared_ptr, как мне кажется, т.к. он используется только в тестирующем коде.

```
void test_unique()
{
    std::vector<std::unique_ptr<Message>> expected = {
        std::unique_ptr<Message>(new MessageError(MessageError::ErrorType::NoSchedule)),
    };

    std::vector<std::unique_ptr<Message>> actual = {
        std::unique_ptr<Message>(new MessageError(MessageError::ErrorType::NoTimestamp)),
    };

    compareVectorsOfSharedPtrs(expected, actual);
}
```

```/usr/include/c++/11/bits/stl_uninitialized.h: In instantiation of ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = const std::unique_ptr<Message>*; _ForwardIterator = std::unique_ptr<Message>*]’:
/usr/include/c++/11/bits/stl_uninitialized.h:333:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = const std::unique_ptr<Message>*; _ForwardIterator = std::unique_ptr<Message>*; _Tp = std::unique_ptr<Message>]’
/usr/include/c++/11/bits/stl_vector.h:1585:33:   required from ‘void std::vector<_Tp, _Alloc>::_M_range_initialize(_ForwardIterator, _ForwardIterator, std::forward_iterator_tag) [with _ForwardIterator = const std::unique_ptr<Message>*; _Tp = std::unique_ptr<Message>; _Alloc = std::allocator<std::unique_ptr<Message> >]’
/usr/include/c++/11/bits/stl_vector.h:629:21:   required from ‘std::vector<_Tp, _Alloc>::vector(std::initializer_list<_Tp>, const allocator_type&) [with _Tp = std::unique_ptr<Message>; _Alloc = std::allocator<std::unique_ptr<Message> >; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<std::unique_ptr<Message> >]’
/home/a/prosoftschool2023/devicemonitoringserver/tests.cpp:813:5:   required from here
/usr/include/c++/11/bits/stl_uninitialized.h:138:72: error: static assertion failed: result type must be constructible from value type of input range
  138 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
      |                                                                        ^~~~~
/usr/include/c++/11/bits/stl_uninitialized.h:138:72: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
```

anzotov replied:
Хм, заменил shared_ptr на unique_ptr только в этом классе, все компилируется и работает. Значит действительно, здесь shared_ptr ни к чему. А вот в тестах его так просто не заменить, получается.

anzotov replied:
Сначала думал, что код с unique_ptr не будет компилироваться, но теперь разобрался и понял, что компилироваться будет.

anzotov replied:
Выполнено.

anzotov committed:
Исправление замечаний после ревью

anzotov committed:
Убран ненужный header, расширен тест сериализации

anzotov committed:
Устранение замечаний

Все override объявлены final
Коллбеки в CommandCenter заменены на unique_ptr
Коллбеки при сериализации заменены на unique_ptr
Методы сериализации сделаны статическими
Переделка сериализации
